plugins/chromatic_correction.py

```diff
@@ -0,0 +1,355 @@
+        unique_elements = set(map(tuple, all_emission))
+        check_unique_channels = len(unique_elements) == len(all_emission)
+        self.dlg = AcquisitionDialog(self, "Chromatic Correction",
+                                     ("Ready to start Chromatic aberration on the given channel.\n"
+                                      "The process will take about 5 minutes\n\n"
+                                      "Z stack range is {} m\n"
+                                      "Z step size of {} m".format(Z_RANGE, Z_STEP_SIZE)))
+        if (len(tab_data.streams.value) <= 1) or (not check_ref_channel) or (not check_unique_channels):
+            txt = ("Problems\n1. Add minimum of two streams.\n2. One of the streams must be green channel.\n"
+                   "3. No two streams should be the same.")
+            self.dlg.lbl_description.SetLabel(txt)
+            self.dlg.addButton("Manually fix")
+            self.dlg.ShowModal()
+            return
+        self.dlg.Bind(wx.EVT_CLOSE, self.on_close)
+        self.dlg.addButton("Start", self._on_start, face_colour='blue')
+        self.dlg.ShowModal()
+
+    def on_close(self, evt, *args, **kwargs):
+        self.dlg.buttons[0].SetLabel("Cancelling...")
+        self.dlg.buttons[0].Disable()
+        self.dlg.lbl_description.SetLabel("Cancellation in progress\nMay take some time to "
+                                          "cancel the already running processes")
+        self.dlg.Destroy()
+        if self.acqui_task:
+            self.acqui_task.cancel()
+        logging.debug("Dialog closed")
+
+    def _on_start(self, evt):
+        """Start or cancel the correction method when the button is clicked"""
+        # If correction is running, cancel it, otherwise start one
+        if self.dlg.buttons[0].Label == "Start":
+            self.cancelled = False
+            self._start_chromatic_correction()
+        else:
+            self.on_close(evt)
+
+    def _start_chromatic_correction(self):
+        # Change the Start button to Cancel
+        self.dlg.buttons[0].SetLabel("Cancel")
+        localization_tab = self.main_app.main_data.getTabByName("cryosecom-localization")
+        tab_data = localization_tab.tab_data_model
+        self.dlg.lbl_description.SetLabel("In Progress\nAcquiring Z stack of the given channels")
+        mip_das_dict = {}  # keys are filter band positions and values are mip data arrays
+        # Acquire the Maximum Intensity Projection for the given streams
+        levels = generate_zlevels(tab_data.main.focus, Z_RANGE, Z_STEP_SIZE)
+        streams_list = [s for s in tab_data.streams.value
+                        if isinstance(s, FluoStream)]
+        zlevels = {stream: levels for stream in streams_list}
+        settings_observer = tab_data.main.settings_obs
+        self.acqui_task = acquireZStack(streams_list, zlevels, settings_obs=settings_observer)
+
+        try:
+            da_list, _ = self.acqui_task.result()
```

Comment:
Instead of blocking until the end of the acquisition, you can start this task from the GUI thread, and then `add_done_callback()` to call a separate function that will handle the end of the task (either completed, or cancelled).

Reply:
I have added the callback and decorated it.